src/literal.h

```diff
@@ -33,52 +33,69 @@ class Literals;
 struct ExceptionPackage;
 
 class Literal {
+public:
+  // Type of the literal. Immutable because the literal's payload depends on it.
+  const Type type;
+
+private:
   // store only integers, whose bits are deterministic. floats
   // can have their signalling bit set, for example.
   union {
+    // i32 or f32 bits
     int32_t i32;
+    // i64 or f64 bits
     int64_t i64;
+    // v128 bytes
     uint8_t v128[16];
-    Name func; // function name for funcref
-    std::unique_ptr<ExceptionPackage> exn;
+    // funcref function name. not set indicates `null`.
+    Name func;
+    // exnref package. `nullptr` indicates `null`. we use a pointer here to
+    // limit storage size, with the lifetime of the data managed by the literal.
+    ExceptionPackage* exn;
+    // literals of other reference types can only be `null`
   };
 
 public:
-  Type type;
-
-public:
-  Literal() : v128(), type(Type::none) {}
-  explicit Literal(Type type) : v128(), type(type) {
-    assert(type != Type::unreachable);
+  Literal() : type(Type::none), v128() {}
+  explicit Literal(Type::BasicID typeId) : Literal(Type(typeId)) {}
```

Comment:
With the type being immutable came the need to construct literals as `Literal(Type::f32)` for example, but that picked the `uint32_t` overload wrongly making it an integer, so I added this one as well.